cw/util/trace_library.py

```diff
@@ -0,0 +1,231 @@
+        """
+        if self.trace_mem:
+            query = db.insert(self.traces_table)
+            traces = []
+            for trace in self.trace_mem:
+                tr = asdict(trace)
+                del tr["trace_id"]
+                traces.append(tr)
+            self.session.execute(query, traces)
+            self.session.commit()
+            self.trace_mem = []
+
+    def write_to_buffer(self, trace):
+        """ Write traces into memory or into storage.
+
+        The trace first is written into memory. When reaching the trace
+        memory threshold, the traces are flushed into the database storage.
+
+        Args:
+            trace: The trace to write.
+        """
+        self.trace_mem.append(trace)
+        if (len(self.trace_mem) >= self.trace_mem_thr):
+            self.flush_to_disk()
+
+    def get_traces(self, start=0, end=0):
+        """ Get traces from database and stored into RAM.
+
+        Fetch traces from start to end from database storage into RAM.
+
+        Args:
+            start: fetch traces from trace index start
+            end: to trace index end. If no start and end is provided, all
+                 all traces are returned.
+        Returns:
+            The traces from the database.
+        """
+        self.flush_to_disk()
+
+        if end == 0:
+            query = db.select(self.traces_table)
+        else:
+            query = db.select(self.traces_table).where(
+                (self.traces_table.c.trace_id >= start) &
+                (self.traces_table.c.trace_id <= end))
+
+        ResultProxy = self.session.execute(query)
+
+        traces = []
+        for trace in ResultProxy.fetchall():
+            traces.append(Trace(**trace._mapping))
+        return traces
+
+    def get_waves_bytearray(self):
+        """ Get all waves from the database.
+
+        Returns:
+            The bytearray waves from the database.
+        """
+        self.flush_to_disk()
+        traces = self.get_traces()
+        waves = []
+        for trace in traces:
+            waves.append(trace.wave)
+        return waves
+
+    def get_waves(self):
+        """ Get all waves from the database in the trace array format.
+
+        Returns:
+            The waves from the database in the type wave_datatype.
+        """
+        self.flush_to_disk()
+        traces = self.get_traces()
+        waves = []
+        for trace in traces:
+            waves.append(np.frombuffer(trace.wave, self.wave_datatype))
+        return waves
+
+    def get_plaintexts_int(self, start=0, end=0):
+        """ Get all plaintexts between start and end from the database in the
+        int8 array format.
+
+        Returns:
+            The int plaintexts from the database.
+        """
+        self.flush_to_disk()
+        traces = self.get_traces(start, end)
+        ptx = []
+        for trace in traces:
+            ptx.append(np.frombuffer(trace.plaintext, np.uint8))
+        return ptx
```

Comment:
I think this could equivalently be:
```python
        return [np.frombuffer(trace.plaintext, np.uint8)
                for trace in self.get_traces(start, end)]
```

Similarly for `get_keys_int`.